File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/CountingSort.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (left > right ? left : right);
*/
static int32_t java_max(int32_t left, int32_t right) {
    return left > right ? left : right;
}

/*@
  assigns \nothing;
  ensures \result == (left < right ? left : right);
*/
static int32_t java_min(int32_t left, int32_t right) {
    return left < right ? left : right;
}

/*@
  logic integer fb_minval{L}(JIntArray a, integer k) = k <= 1 ? a->data[0] : \min(fb_minval(a, k - 1), a->data[k - 1]);
  logic integer fb_maxval{L}(JIntArray a, integer k) = k <= 1 ? a->data[0] : \max(fb_maxval(a, k - 1), a->data[k - 1]);
  logic integer fb_occ{L}(JIntArray a, integer k, integer v) = k <= 0 ? 0 : fb_occ(a, k - 1, v) + (a->data[k - 1] == v ? 1 : 0);
  logic integer fb_less{L}(JIntArray a, integer k, integer v) = k <= 0 ? 0 : fb_less(a, k - 1, v) + (a->data[k - 1] < v ? 1 : 0);
*/
/*@
  requires jintarray_valid_nonnull(myArray) && (myArray->length == 0 || fb_maxval(myArray, myArray->length) - fb_minval(myArray, myArray->length) < INT32_MAX);
  assigns \nothing;
  allocates \result, \result->data;
  frees \nothing;
  ensures jintarray_valid_nonnull(\result);
  ensures \result->length == myArray->length;
  ensures \fresh(\result, sizeof(*\result));
  ensures \result->length > 0 ==> \fresh(\result->data, \result->length * sizeof(int32_t));
  
  ensures jintarray_valid_nonnull(\result);
  ensures \forall integer p, q; 0 <= p < q < \result->length ==> \result->data[p] <= \result->data[q];
  ensures \forall integer v; fb_occ(\result, \result->length, v) == fb_occ{Pre}(myArray, myArray->length, v);
*/
JIntArray countingSort(JIntArray myArray) {
    int32_t length = jarray_length(myArray);
    if (length == 0) {
        return jarray_new(0);
    }

    int32_t max = jarray_get(myArray, 0);
    int32_t min = jarray_get(myArray, 0);
    /*@
      loop invariant 0 <= i <= length && length == myArray->length;
      loop invariant min == fb_minval(myArray, i > 0 ? i : 1);
      loop invariant max == fb_maxval(myArray, i > 0 ? i : 1);
      loop assigns i, min, max;
      loop variant length - i;
    */
    for (int32_t i = 0; i < length; i = java_add(i, 1)) {
        int32_t num = jarray_get(myArray, i);
        max = java_max(max, num);
        min = java_min(min, num);
    }

    int32_t range = java_add(java_sub(max, min), 1);
    JIntArray countArray = jarray_new(range);
    /*@
      loop invariant 0 <= i <= length;
      loop invariant jintarray_valid_nonnull(countArray) && countArray->length == range;
      loop invariant \forall integer v; 0 <= v < range ==> countArray->data[v] == fb_occ(myArray, i, (integer)min + v);
      loop assigns i, countArray->data[0 .. range - 1];
      loop variant length - i;
    */
    for (int32_t i = 0; i < length; i = java_add(i, 1)) {
        int32_t slot = java_sub(jarray_get(myArray, i), min);
        int32_t old_count = jarray_get(countArray, slot);
        jarray_set(countArray, slot, java_add(old_count, 1));
    }

    int32_t index = 0;
    JIntArray result = jarray_new(length);
    int32_t countLength = jarray_length(countArray);
    /*@
      loop invariant 0 <= i <= countLength && countLength == range && 0 <= index <= length;
      loop invariant jintarray_valid_nonnull(result) && result->length == length;
      loop invariant index == fb_less{Pre}(myArray, length, (integer)min + i);
      loop invariant \forall integer p, q; 0 <= p < q < index ==> result->data[p] <= result->data[q];
      loop invariant \forall integer v; fb_occ(result, index, v) == (v < (integer)min + i ? fb_occ{Pre}(myArray, length, v) : 0);
      loop assigns i, index, result->data[0 .. length - 1];
      loop variant countLength - i;
    */
    for (int32_t i = 0; i < countLength; i = java_add(i, 1)) {
        /*@
          loop invariant 0 <= j <= countArray->data[i] && 0 <= index <= length;
          loop invariant jintarray_valid_nonnull(result) && result->length == length;
          loop invariant index == fb_less{Pre}(myArray, length, (integer)min + i) + j;
          loop invariant \forall integer p, q; 0 <= p < q < index ==> result->data[p] <= result->data[q];
          loop invariant \forall integer v; fb_occ(result, index, v) == (v < (integer)min + i ? fb_occ{Pre}(myArray, length, v) : v == (integer)min + i ? j : 0);
          loop assigns j, index, result->data[0 .. length - 1];
          loop variant countArray->data[i] - j;
        */
        for (int32_t j = 0; j < jarray_get(countArray, i); j = java_add(j, 1)) {
            jarray_set(result, index, java_add(i, min));
            index = java_add(index, 1);
        }
    }
    return result;
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/CountingSort.java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;

class CountingSort {
    
    /*@ public normal_behavior
      @ ensures \result == (a <= b ? a : b);
      @ model public static pure \bigint fb_min(\bigint a, \bigint b);
      @*/
    /*@ public normal_behavior
      @ ensures \result == (a >= b ? a : b);
      @ model public static pure \bigint fb_max(\bigint a, \bigint b);
      @*/
    /*@ public normal_behavior
      @ requires a != null && 1 <= k <= a.length;
      @ ensures \result == (k == 1 ? a[0] : fb_min(fb_minval(a, k - 1), a[k - 1]));
      @ measured_by k;
      @ model public static pure \bigint fb_minval(int[] a, int k);
      @*/
    /*@ public normal_behavior
      @ requires a != null && 1 <= k <= a.length;
      @ ensures \result == (k == 1 ? a[0] : fb_max(fb_maxval(a, k - 1), a[k - 1]));
      @ measured_by k;
      @ model public static pure \bigint fb_maxval(int[] a, int k);
      @*/
    /*@ public normal_behavior
      @ requires myArray != null && (myArray.length == 0 || fb_maxval(myArray, myArray.length) - fb_minval(myArray, myArray.length) < Integer.MAX_VALUE);
      @ assignable \nothing;
      @ ensures \result != null && \fresh(\result) && \result.length == myArray.length;
      @ ensures (\forall int p, q; 0 <= p < q < \result.length; \result[p] <= \result[q]);
      @ ensures (\forall int v; (\num_of int k; 0 <= k < \result.length; \result[k] == v) == \old(\num_of int k; 0 <= k < myArray.length; myArray[k] == v));
      @ also
      @ public exceptional_behavior
      @ requires myArray == null;
      @ assignable \nothing;
      @ signals_only NullPointerException;
      @ also
      @ public exceptional_behavior
      @ requires myArray != null && myArray.length > 0 && Integer.MAX_VALUE <= fb_maxval(myArray, myArray.length) - fb_minval(myArray, myArray.length) && fb_maxval(myArray, myArray.length) - fb_minval(myArray, myArray.length) < 4294967295L;
      @ assignable \nothing;
      @ signals_only NegativeArraySizeException;
      @ also
      @ public exceptional_behavior
      @ requires myArray != null && myArray.length > 0 && fb_maxval(myArray, myArray.length) - fb_minval(myArray, myArray.length) == 4294967295L;
      @ assignable \nothing;
      @ signals_only ArrayIndexOutOfBoundsException;
      @*/
    //@ code_java_math
    public static int[] countingSort(int[] myArray) {
        if (myArray.length == 0) {
            return new int[0]; // return an empty array if input is empty
        }
        
        int max = myArray[0];
        int min = myArray[0];

        /*@ loop_invariant 0 <= \count <= myArray.length;
          @ loop_invariant min == fb_minval(myArray, \count > 0 ? \count : 1);
          @ loop_invariant max == fb_maxval(myArray, \count > 0 ? \count : 1);
          @ loop_writes min, max;
          @ decreases myArray.length - \count;
          @*/
        for (int num : myArray) {
            if (num > max) max = num;
            if (num < min) min = num;
        }

        int range = max - min + 1;
        int[] countArray = new int[range];

        /*@ loop_invariant 0 <= i <= myArray.length;
          @ loop_invariant (\forall int v; 0 <= v < countArray.length; countArray[v] == (\num_of int k; 0 <= k < i; myArray[k] == (\bigint)min + v));
          @ loop_writes countArray[*], i;
          @ decreases myArray.length - i;
          @*/
        for (int i = 0; i < myArray.length; i++) {
            countArray[myArray[i] - min]++;
        }

        int index = 0;
        int[] result = new int[myArray.length];

        /*@ loop_invariant 0 <= i <= countArray.length && 0 <= index <= myArray.length;
          @ loop_invariant index == ((\num_of int k; 0 <= k < myArray.length; \old(myArray[k]) < (\bigint)min + i));
          @ loop_invariant (\forall int p, q; 0 <= p < q < index; result[p] <= result[q]);
          @ loop_invariant (\forall int v; (\num_of int k; 0 <= k < index; result[k] == v) == (v < (\bigint)min + i ? (\num_of int k; 0 <= k < myArray.length; \old(myArray[k]) == v) : 0));
          @ loop_writes i, index, result[*];
          @ decreases countArray.length - i;
          @*/
        for (int i = 0; i < countArray.length; i++) {
            /*@ loop_invariant 0 <= j <= countArray[i] && 0 <= index <= myArray.length;
              @ loop_invariant index == ((\num_of int k; 0 <= k < myArray.length; \old(myArray[k]) < (\bigint)min + i)) + j;
              @ loop_invariant (\forall int p, q; 0 <= p < q < index; result[p] <= result[q]);
              @ loop_invariant (\forall int v; (\num_of int k; 0 <= k < index; result[k] == v) == (v < (\bigint)min + i ? (\num_of int k; 0 <= k < myArray.length; \old(myArray[k]) == v) : v == (\bigint)min + i ? j : 0));
              @ loop_writes j, index, result[*];
              @ decreases countArray[i] - j;
              @*/
            for (int j = 0; j < countArray[i]; j++) {
                result[index++] = i + min;
            }
        }
        return result;
    }
}

